master/src/main/java/org/apache/celeborn/service/deploy/master/clustermeta/AbstractMetaManager.java

```diff
@@ -284,27 +289,49 @@ public void updateWorkerHeartbeatMeta(
     }
 
     // If using HDFSONLY mode, workers with empty disks should not be put into excluded worker list.
-    long unhealthyDiskNum =
-        disks.values().stream().filter(s -> !s.status().equals(DiskStatus.HEALTHY)).count();
-    boolean exceed = unhealthyDiskNum * 1.0 / disks.size() >= unhealthyDiskRatioThreshold;
+    Pair<Boolean, Long> exceedCheckResult = isExceedingUnhealthyThreshold(disks);
     if (!excludedWorkers.contains(worker)
-        && (((disks.isEmpty() || exceed)
-                && !conf.hasHDFSStorage()
-                && !conf.hasS3Storage()
-                && !conf.hasOssStorage())
+        && (((disks.isEmpty() || exceedCheckResult.getLeft()) && !hasRemoteStorage())
             || highWorkload)) {
       LOG.warn(
-          "Worker {} (unhealthy disks num: {}) adds to excluded workers", worker, unhealthyDiskNum);
+          "Worker {} (unhealthy disks num: {}) adds to excluded workers",
+          worker,
+          exceedCheckResult.getRight());
       excludedWorkers.add(worker);
-    } else if ((availableSlots.get() > 0
-            || conf.hasHDFSStorage()
-            || conf.hasS3Storage()
-            || conf.hasOssStorage())
-        && !highWorkload) {
+    } else if ((availableSlots.get() > 0 || hasRemoteStorage()) && !highWorkload) {
       // only unblack if numSlots larger than 0
       excludedWorkers.remove(worker);
     }
 
+    // release high work load workers when too many excluded workers
+    if (releaseHighWorkLoadEnabled
+        && excludedWorkers.size()
+            >= Math.floor(workersMap.size() * releaseHighWorkLoadRatioThreshold)) {
+      synchronized (workersMap) {
+        List<WorkerInfo> toRemoved =
+            excludedWorkers.stream()
+                .filter(
+                    w -> {
+                      Optional<WorkerInfo> infoOpt =
+                          Optional.ofNullable(workersMap.get(w.toUniqueId()));
+                      if (infoOpt.isPresent()) {
+                        WorkerInfo info = infoOpt.get();
+                        Map<String, DiskInfo> diskInfos = info.diskInfos();
+                        if (info.isHighWorkLoad()
+                            && (hasRemoteStorage()
```

Comment:
should be `hasRemoteStorage()
|| (!diskInfos.isEmpty()
&& !isExceedingUnhealthyThreshold(diskInfos).getLeft())`, we can also use `hasStorage()`?